### PR TITLE
Safely enforce thread name requirements

### PR DESCRIPTION
**Pull request by ChrisDenton**
The requirements for the thread name to be both UTF-8 and null terminated are easily enforced by a wrapper type so lets do that. The fact this used to be just a bare `CString` has tripped me up before because it was entirely safe to use a non UTF-8 `CString`.

### CODE REVIEW

File: library/std/src/thread/mod.rs [hunk 1]
```diff
@@ -1316,10 +1336,7 @@ pub struct Thread {
 
 impl Thread {
     /// Used only internally to construct a thread object without spawning.
-    ///
-    /// # Safety
-    /// `name` must be valid UTF-8.
-    pub(crate) unsafe fn new(name: CString) -> Thread {
+    pub(crate) fn new(name: ThreadNameString) -> Thread {
         unsafe { Self::new_inner(ThreadName::Other(name)) }
```
Comment (joboet) [inline library/std/src/thread/mod.rs:1340]: This makes the `unsafe` on `new_inner` redundant.

Comment (ChrisDenton) [reply library/std/src/thread/mod.rs:1340]: Ah, good point

File: library/std/src/thread/mod.rs [hunk 2]
```diff
@@ -1316,25 +1336,20 @@ pub struct Thread {
...
 impl Thread {
     /// Used only internally to construct a thread object without spawning.
-    ///
-    /// # Safety
-    /// `name` must be valid UTF-8.
-    pub(crate) unsafe fn new(name: CString) -> Thread {
-        unsafe { Self::new_inner(ThreadName::Other(name)) }
+    pub(crate) fn new(name: ThreadNameString) -> Thread {
+        Self::new_inner(ThreadName::Other(name))
     }
```
Comment (joboet) [inline library/std/src/thread/mod.rs:1341]:
```suggestion
    pub(crate) fn new(name: String) -> Thread {
        Self::new_inner(ThreadName::Other(name.into()))
    }
```

File: library/std/src/thread/mod.rs [hunk 3]
```diff
@@ -487,11 +487,7 @@ impl Builder {
             amt
         });
 
-        let my_thread = name.map_or_else(Thread::new_unnamed, |name| unsafe {
-            Thread::new(
-                CString::new(name).expect("thread name may not contain interior null bytes"),
-            )
-        });
+        let my_thread = name.map_or_else(Thread::new_unnamed, |name| Thread::new(name.into()));
```
Comment (joboet) [inline library/std/src/thread/mod.rs:490]:
```suggestion
        let my_thread = name.map_or_else(Thread::new_unnamed, Thread::new);
```

**Review by joboet — CHANGES_REQUESTED** [review]
One style recommendation, I don't insist on it.

On this I'll insist however: could you move the `from_utf8_unchecked` call from `Thread::name` into the module? That way, the safety invariant is local.

r? @joboet

**Comment by ChrisDenton** [comment]
I moved both the `&cstr` and `&str` conversion into the module to keep everything right next to each other.

**Pull request merged**